### PR TITLE
Revamp QA checks into a battery included package

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Relates to:
* [#2849](https://github.com/airbytehq/airbyte-internal-issues/issues/2849)
* [#2905](https://github.com/airbytehq/airbyte-internal-issues/issues/2905)
* [#2908](https://github.com/airbytehq/airbyte-internal-issues/issues/2908)
* [#2907](https://github.com/airbytehq/airbyte-internal-issues/issues/2907)

This PR introduces a new `connectors-qa` 🐍  package which can run static-analysis checks on our connectors and generate documentation.

It will help address the following problems we have:

* Our connectors standards are poorly documented
* Our QA checks (enforcing connector standards) are scattered in multiple packages: 
   * The [`qa_checks.py` script](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/connector_ops/connector_ops/qa_checks.py#L1) in the `connector_ops` package (called by `airbyte-ci`)
   * In `airbyte-ci` (e.g.[ `VersionFollowsSemverCheck` ](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py#L122)or [`CheckPythonRegistryPublishConfiguration`](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/test/steps/common.py#L347))
  
## 🤔 Philosophy
This efforts is driven by the following principles:
* **Our connector standards apply to all connectors:**
    * Airbyte should commit that certified connectors will always meet all the standard.
    * Any new contribution on a community connector must meet all standards to be merged. 
* We should limit as much as possible any connector specific logic in`airbyte-ci`. We should consider it an orchestrator calling external tools and packages. `airbyte-ci` will call  a containerized`connectors-qa` via its CLI.

### Documentation generation
```
connectors-qa generate-documentation connectors_qa_documentation.md
```

This will generate a [markdown file](https://gist.github.com/alafanechere/5ea65d8c7890685506cf54a0cd22e4df)documenting all the enabled checks.
The content is taken from check classes names and descriptions.

### Report generation
```bash
connectors-qa run --connector-directory=airbyte-integrations/connectors --report-path=qa_report.json
```
This will generate [a json report ](https://gist.githubusercontent.com/alafanechere/a4af369a6e41ee32d0b7f87fdfda8f40/raw/2cdac6ebeaba239d5c4ba2517b5a199c06b75514/qa_report.json) of all the QA checks on all our connectors.
We could automate the generation of this report in our CD pipeline and use it to feed dashboards or other assets like connector's README.
And create cool badges like: Dynamic JSON Badge

## Recommended reading order
1. `README.md` to install and try out the tool locally.
2. `checks/*.py` to understand which checks are running
3. `cli.py` to grasp how the entrypoint and `asyncio` logic is implemented

## 🚨 User Impact 🚨
None as this PR is just introducing a new package which is not yet used by `airbyte-ci`

## Follow up steps
0. Commit the generated documentation https://github.com/airbytehq/airbyte/pull/35324
1. Call this package in the `QAChecks` step of `airbyte-ci connectors test`
2. Remove `connector_ops/qa_checks` and `airbyte-ci` steps that are now running inside this package (semver version check, pypi publishing etc.)
4. Automate the report generation and host it on GCS to power dashboards - connector readme badges?
5. Replace the "Connector checklist" on PR by a comment linking to the generate doc?

### CODE REVIEW

File: airbyte-ci/connectors/connector_ops/pyproject.toml
Comment (bazarnov) [inline airbyte-ci/connectors/connector_ops/pyproject.toml:17]: What holds our back to switch to the pydantic 2.0.0? Just curious?

File: airbyte-ci/connectors/connectors_qa/connectors_qa/checks/documentation.py
Comment (bazarnov) [inline airbyte-ci/connectors/connectors_qa/connectors_qa/checks/documentation.py:38]: Do we have all the critical rules covered?

**Review by bazarnov — COMMENTED** [review]
Looks very promising to me, thanks @alafanechere ! Not approving now, because i'll get a closer look tomorrow to see the big picture of the checks themselves.

File: airbyte-ci/connectors/connector_ops/pyproject.toml
Comment (alafanechere) [reply airbyte-ci/connectors/connector_ops/pyproject.toml:17]: I don't know, nothing I'm aware of, but it's a different package / project than the current one :)

Comment (natikgadzhi) [reply airbyte-ci/connectors/connector_ops/pyproject.toml:17]: I don't yet know the full surface area of us using Pydantic. If we're using 1.* everywhere, should we treat 2.0+ upgrade as a separate lane of work, or do you mostly expect things to work smoothly with such an upgrade?

Is is drop-in compatible syntax-wise?

If we're using it in the CDK itself and in airbyte-ci — I think airbyte-ci could be our guinea pig for such a migration? /cc @bazarnov

File: airbyte-ci/connectors/connector_ops/pyproject.toml
Comment (natikgadzhi) [inline airbyte-ci/connectors/connector_ops/pyproject.toml:20]: Why downgrade?

File: airbyte-ci/connectors/connectors_qa/README.md
Comment (natikgadzhi) [inline airbyte-ci/connectors/connectors_qa/README.md:1]: Two questions here:

1. Does `connectors_qa` have to live in this particular directory, or is it just logical spot for it in the monorepo? I.e. could we just put it in the root of the repository instead?
2. If it's fully independent of the rest of connector_ops stuff (don't think so), is there a good overall readme and list of directories and what project is where? We have many things — registry service, airbyte-ci, etc.

File: airbyte-ci/connectors/connectors_qa/connectors_qa/checks/__init__.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/connectors_qa/connectors_qa/checks/__init__.py:9]: I was about to suggest that you could have a list of lists and then flatten, but then I remembered how flattening lists in Python is expressed kek.

File: airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py:11]: How is Gradle used internally, and for what?

File: airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py
Comment (natikgadzhi) [inline airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py:10]: Wow wow wow hold up, why are our standards in a 3rd party service and not a page on our very own docs site? Any specific reason? /cc @girarda @alafanechere

**Review by natikgadzhi — APPROVED** [review]
This does look great to me — since the potential blast radius of just _having_ the tool is close to zero, I'm comfy approving this. 

Caveats: I want follow-up work of:
- Publishing the actual generated documentation
- Moving our doc guide to our doc pages instead of hackmd
- Actually switching to use `connectors-qa`. 

Please wait for @bazarnov's review and work with him to get a go ahead from API Sources, but I'm personally happy.

File: airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py
Comment (girarda) [reply airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py:10]: No idea. This is the first time I see this page. I think this was hacked together by dev-rel two years ago, but that's only based off this [slack thread](https://airbytehq-team.slack.com/archives/C02UQ9MJ4GG/p1653835061244979).

File: airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py
Comment (alafanechere) [reply airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py:11]: @natikgadzhi these checks run also on our java connectors which are using gradle. 
This file is used in this program to determine if a connector is a java one.

File: airbyte-ci/connectors/connectors_qa/README.md
Comment (alafanechere) [reply airbyte-ci/connectors/connectors_qa/README.md:1]: @natikgadzhi I put it there because it's a package used in the CI context. But we could put it at the root of the repo, I don't mind 😄 . 
It's dependent on a couple of other internal packages. I will explicitly list the dependency and why they exists in this README.

File: airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py
Comment (alafanechere) [reply airbyte-ci/connectors/connectors_qa/connectors_qa/consts.py:10]: @natikgadzhi @girarda this doc was written by our previous documentation team. I believe it was drafted when I was working on the original `qa_checks.py` and never got checked in our main repo. I believe the doc is good enough to join our `Resources` section of `Contributing to Airbyte`. I can do it in a follow up PR.

Screenshot 2024-02-16 at 08 09 30

**Comment by alafanechere** [comment]
> This does look great to me — since the potential blast radius of just _having_ the tool is close to zero, I'm comfy approving this.
> 
> Caveats: I want follow-up work of:
> 
> * Publishing the actual generated documentation
> * Moving our doc guide to our doc pages instead of hackmd
> * Actually switching to use `connectors-qa`.
> 
> Please wait for @bazarnov's review and work with him to get a go ahead from API Sources, but I'm personally happy.

@natikgadzhi I generated the documentation + add an integration test to make sure it's kept up to date in https://github.com/airbytehq/airbyte/pull/35324

File: airbyte-ci/connectors/connector_ops/pyproject.toml
Comment (alafanechere) [reply airbyte-ci/connectors/connector_ops/pyproject.toml:20]: `connectors_qa` depends on two other internal packages which are using `pydash` but on different version:
* `medata_service/lib` uses `6`
* `connector_ops` uses `7`

 I'm aligning to  `metadata_service/lib` version because this package is also a dependency of `metadata_service/orchestrator`, which is also declaring a dependency on `pydash`.

Aligning to `6` makes me modifying a single package (`connector_ops`) instead of 2 (`metadata_service/lib` / `metadata_service/orchestrator`). A part from that I don't think there's a good reason to stay on `6`.

File: airbyte-ci/connectors/connectors_qa/connectors_qa/checks/documentation.py
Comment (alafanechere) [reply airbyte-ci/connectors/connectors_qa/connectors_qa/checks/documentation.py:38]: I ported over what currently exists in [`qa_checks.py` ](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/connector_ops/connector_ops/qa_checks.py). Let me know if you think additional checks should be implemented. I'm open for it, but in different PRs and with related new GH issues 😄

**Review by bazarnov — APPROVED** [review]
LGTM. Thanks @alafanechere for this improvement.

**Pull request merged**
